refactor: migrate to Yarn Berry (v3.2.1)

ThatOneCalculator opened the pull request:
# What

Migrate to Yarn Berry (v3.2.1)

Resolve #5858 
Supersede #8472 

# Why

As yarn v3 (yarn berry) has worked with Misskey for a while and provides better performance when compiling while still retaining backwards compatibility with yarn v1, I feel that upgrading to yarn berry would be a smart choice. 

# Additional info

I've been building Misskey with yarn v3 for two months now both on my server for production and on my development machine with absolutely no problems. This PR was made simply by running `yarn` with yarn 3.2.1 and following the [yarn migration instructions](https://yarnpkg.com/getting-started/migration).

# Further steps

- [ ] Documentation on Misskey Hub would have to be updated
- [x] Update workflow (especially mocha and e2e)
- [x] Update the install script (maybe docker too?)

ThatOneCalculator committed:
Yarn berry

ThatOneCalculator committed:
Add `.yarn` to dockerignore

ThatOneCalculator committed:
Mention in CHANGELOG

ThatOneCalculator committed:
Update Dockerfile

ThatOneCalculator committed:
Proper upgrade to yarn

ThatOneCalculator commented:
@acid-chicken I'd personally like your feedback on this PR first, as you've been the most involved with Yarn Berry discussion.

acid-chicken reviewed (commented):
Thank you for your work.
I believe that by using the node_modules linker, we can upgrade without facing most of the problems, so I think this PR is fine with the current policy for now, regardless of whether we move to PnP mode in the future or not.

Also, since Yarn berry has built-in monorepo management system called workspaces, so it would be great (and probably easy) to also improve the current situation, which is awkwardly realized by `scripts/install-packages.js` and so on.

acid-chicken commented:
Also, since there are opinions like https://github.com/misskey-dev/misskey/issues/5858#issuecomment-611626477, it would be better if Corepack could be used as a boot loader for Yarn berry.

ThatOneCalculator commented:
Running `corepack enable` in the repo folder didn't change anything, is there a proper way to enable corepack?

acid-chicken commented:
`corepack enable` is related to environment settings, so it will not change anything in the repository (excluding docs).
Can you confirm that deleting `.yarn/releases/yarn-3.2.1.cjs` with ["packageManager"](https://nodejs.org/api/packages.html#packagemanager) specified in package.json after running it still works?

ThatOneCalculator committed:
use `"packageManager"`

ThatOneCalculator commented:
> Can you confirm that deleting .yarn/releases/yarn-3.2.1.cjs with ["packageManager"](https://nodejs.org/api/packages.html#packagemanager) specified in package.json after running it still works?

@acid-chicken yep!

ThatOneCalculator commented:
Seems like the labeler bot adds back removed labels every commit ^^"

ThatOneCalculator committed:
Replace `install-packages.js` with workspaces

ThatOneCalculator committed:
Replace `install-packages.js` with workspaces

ThatOneCalculator commented:
> Also, since Yarn berry has built-in monorepo management system called workspaces, so it would be great (and probably easy) to also improve the current situation, which is awkwardly realized by scripts/install-packages.js and so on.

Done!

ThatOneCalculator committed:
Typo

acid-chicken commented:
I assume `packages/*/yarn.lock` will no longer be used after `yarn install` should consolidate the lock files in `/yarn.lock`.

ThatOneCalculator commented:
Should I add `packages/*/yarn.lock` to `.gitignore`?

ThatOneCalculator commented:
Also seems like removing `yarn-3.2.1.cjs` broke the linting workflows...

![image](https://user-images.githubusercontent.com/44733677/170899572-9e44cd69-be2a-4e4b-b484-7c6064500827.png)

acid-chicken commented:
*`corepack enable` is related to environment settings*, so, for now, we have run it manually after `setup-node` (actions/setup-node#482).

ThatOneCalculator committed:
`corepack enable` for linting

ThatOneCalculator commented:
Fixed.

ThatOneCalculator commented:
Seems like the current lint fails because of 

https://github.com/ThatOneCalculator/misskey/blob/7df83d72526d2e1147494c61df57dfeaba8fda04/.github/workflows/lint.yml#L21-L22

causing

![image](https://user-images.githubusercontent.com/44733677/170900397-87c8ee94-8c0e-4a1c-9ea1-7c2822c249b7.png)

So hopefully it'll fix itself after this is merged.

acid-chicken commented:
> Should I add `packages/*/yarn.lock` to `.gitignore`?

No. You have to run `yarn install` first, then they can safely deleted.

ThatOneCalculator committed:
Remove `packages/*/yarn.lock`

ThatOneCalculator committed:
Merge branch 'yarn-3' of https://github.com/ThatOneCalculator/misskey into yarn-3

acid-chicken commented:
Also `yarn --cwd ./packages/* ...` are should be rewritten to `yarn workspaces * ...`

https://github.com/misskey-dev/misskey/blob/ebc2566130c785924380d504c76b270198bdc2c2/.github/workflows/lint.yml#L24

ThatOneCalculator commented:
@acid-chicken like that?

ThatOneCalculator commented:
@acid-chicken before I commit, something like this?

```yaml
jobs:
  workspaces:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
      with:
        submodules: true
    - uses: actions/setup-node@v3
      with:
        node-version: 18.x
        cache: 'yarn'
        cache-dependency-path: |
          yarn.lock
    - run: corepack enable
    - run: yarn install
    - run: yarn workspaces * lint
```

acid-chicken commented:
Yes. More to say, it seems that `cache-dependency-path` is no longer needed.

ThatOneCalculator committed:
Improve lint workflow

ThatOneCalculator commented:
@acid-chicken Hmm... `yarn install` still fails in the workflow

<details><summary>YN0028: │ The lockfile would have been modified by this install, which is explicitly forbidden.
</summary>

![image](https://user-images.githubusercontent.com/44733677/170901377-022e29fd-4158-406d-9915-6cc2e1e3e0a2.png)

</details>

acid-chicken commented on package.json:
```diff
@@ -6,9 +6,14 @@
 	},
+	"packageManager": "yarn@3.2.1",
+	"workspaces": [
+		"client",
+		"backend",
+		"sw"
```
Are they works without `packages/`?

ThatOneCalculator replied:
Yea, `workspaces` automatically looks in the `packages` folder.

ThatOneCalculator replied:
After some more research, turns out I was entirely wrong -_-

acid-chicken commented on .github/workflows/lint.yml:
```diff
@@ -8,7 +8,7 @@ on:
 
 jobs:
-  backend:
+  workspaces:
```
```suggestion
  lint:
    strategy:
      matrix:
        workspace:
        - backend
        - client
```

acid-chicken commented on .github/workflows/lint.yml:
```diff
@@ -18,22 +18,6 @@ jobs:
-          packages/client/yarn.lock
-    - run: yarn install
-    - run: yarn --cwd ./packages/client lint
+    - run: yarn workspaces * lint
```
I mean:

```suggestion
    - run: yarn workspaces ${{ matrix.workspace }} lint
```
```

ThatOneCalculator committed:
Update .github/workflows/lint.yml

Co-authored-by: Acid Chicken (硫酸鶏) <root@acid-chicken.com>

ThatOneCalculator committed:
Update .github/workflows/lint.yml

Co-authored-by: Acid Chicken (硫酸鶏) <root@acid-chicken.com>

ThatOneCalculator committed:
Add `eslint`

ThatOneCalculator committed:
Regenerate yarn.lock

ThatOneCalculator commented:
Workflow failure seems to be related to https://github.com/yarnpkg/berry/issues/2948 / https://stackoverflow.com/questions/67062308/getting-yn0028-the-lockfile-would-have-been-modified-by-this-install-which-is-e

ThatOneCalculator committed:
🙏‍

ThatOneCalculator commented:
Nope, gonna try again...

ThatOneCalculator committed:
Update lint.yml

:thinking:

ThatOneCalculator committed:
Merge branch 'misskey-dev:develop' into yarn-3

ThatOneCalculator commented:
@acid-chicken Any ideas? I've been using `act` to test workflows locally but I can't get anything to work :disappointed:

ThatOneCalculator committed:
Fix workflow!

ThatOneCalculator committed:
yarn lint my beloved

ThatOneCalculator commented:
Turns out `yarn workspaces <client/backend> lint` isn't actually a valid command :smile:

ThatOneCalculator commented:
Lint now fails at the expected place (since Misskey still has a lot of lint errors) :D

ThatOneCalculator committed:
corepack enable for mocha/e2e

ThatOneCalculator committed:
Update CHANGELOG.md

ThatOneCalculator commented:
Runs as expected on stop.voring.me, btw

ThatOneCalculator commented:
Runs as expected on stop.voring.me, btw. Suppose I should ask @syuilo for a review as well.

ThatOneCalculator commented:
>  `➤ YN0000: │ /tmp/xfs-88390f66 STDOUT Using Yarn Classic for bootstrap. Reason: "__metadata" key not found in yarn.lock, must be a Yarn classic lockfile`

:confounded: Literally no clue how to fix this

ishowta commented:
node_modulesに依存しているコードを修正する必要があると思います
misskeyのコードベースは小規模ではないのでworkspacesへの移行は大変だと思います

![2022-05-30_133716](https://user-images.githubusercontent.com/7393198/170918047-3765464e-a939-4eb8-b29e-6388898bb0a1.png)

ThatOneCalculator commented:
@ishowta If that much migrating needs to be done, would it be worth it to change to PnP?

ishowta commented:
すみません、yarn pnpを使用したことが無いので分かりません。しかし、移行はより難しくなると思います。例えば依存パッケージ内に存在するcssファイルを参照することは出来るのでしょうか…

良し悪しは分かりませんが、hoistingを無効にして各ディレクトリにnode_modulesを配置するオプションがあります。　https://yarnpkg.com/configuration/yarnrc#nmHoistingLimits

workspacesの機能はyarn v1にも存在するので、berryへのアップグレードとは独立して考えたほうがいいと思います。

Sorry, I don't know, I have never used yarn pnp. But I think the transition will be more difficult. For example, is it possible to reference css files that exist in dependent packages...

I don't know if it is good or bad, but there is an option to disable hoisting and put node_modules in each directory.　https://yarnpkg.com/configuration/yarnrc#nmHoistingLimits

The workspaces feature is also present in yarn v1, so I think it should be considered independent of the upgrade to berry.

ThatOneCalculator committed:
fix the code that depends on node_modules

ThatOneCalculator commented:
@ishowta That should be everything that depends on `node_modules`

ThatOneCalculator commented:
I should add that it builds and runs fine in production on https://stop.voring.me but I'm not sure if I did everything right (plus, workflow is still broken)

acid-chicken commented:
> すみません、yarn pnpを使用したことが無いので分かりません。しかし、移行はより難しくなると思います。例えば依存パッケージ内に存在するcssファイルをcssから参照することは出来るのでしょうか…
>
> 良し悪しは分かりませんが、hoistingを無効にして各ディレクトリにnode_modulesを配置するオプションがあります。　https://yarnpkg.com/configuration/yarnrc#nmHoistingLimits
>
> workspacesの機能はyarn v1にも存在するので、berryへのアップグレードとは独立して考えたほうがいいと思います。

@ishowta https://github.com/misskey-dev/misskey/pull/8764#pullrequestreview-988673362 でも触れましたが、もちろん PnP mode を最初から有効にするのは移行コストが重いので、この PR では node_modules リンカによる移行で良いと思います（し、現状の差分ではそうなっています）。PnP を使用するかどうかと Yarn berry への移行を行うこと自体は別の話です。なお、node_modules パス周りで若干の修正が入っているのは、元々のモノリポ構成が手動で実現されていた点が経緯にあるものと思われます。

ishowta commented:
@acid-chicken あー私も同意見です。node_modulesがhoistingされている場合にパスを修正しなければいけないということを指して「コードを修正する必要があると思います」と言ったつもりでした。私が英語を読み間違えて的はずれな返答をしてしまっていました。すみません。

ThatOneCalculator commented:
Gotcha. With that said, what more should be done on this PR?

ishowta commented:
もしyarn v2のlockfileフォーマットに対応していないパッケージがあれば修正する必要がありますが、それくらいだと思います。（テストが適切に行われていれば）

If there are packages that don't support the yarn v2 lockfile format, they will need to be modified, but I think that's about it. (As long as testing is done properly.)

ThatOneCalculator committed:
tbh

ThatOneCalculator committed:
tbh

ThatOneCalculator committed:
More yarn workspaces stuff

ThatOneCalculator committed:
🍀

ThatOneCalculator commented:
Opening up again for review as build works fine, but workflows are still broken. Help would be greatly appreciated :pray:

ishowta commented:
@ThatOneCalculator ローカルでyarn installをしてみたところCIと同じエラーが発生しました。https://github.com/yarnpkg/berry/issues/1142#issuecomment-1017624835 を参考に`YARN_CHECKSUM_BEHAVIOR=update yarn install`をしたらchecksumが更新されました。yarn v2のバグかもしれません

```diff
diff --git a/yarn.lock b/yarn.lock
index 555881058..406009afd 100644
--- a/yarn.lock
+++ b/yarn.lock
@@ -2884,7 +2884,7 @@ __metadata:
 "browser-image-resizer@git+https://github.com/misskey-dev/browser-image-resizer#v2.2.1-misskey.2":
   version: 2.2.1-misskey.2
   resolution: "browser-image-resizer@https://github.com/misskey-dev/browser-image-resizer.git#commit=a58834f5fe2af9f9f31ff115121aef3de6f9d416"
-  checksum: 5a29181ee34a407b43069dcd5e3364c3066eb4af3ffdd262ba669ea37273ac232aadebff8204a6ae9ff8d74305429191feedb5d420449acf01e1bbf9759c4847
+  checksum: eb5ddfe7f6a2de96340ef420df9be03a75f2bfd8f568a60be22cc8f2ccdcb754105b7799cf706c09add3f9d82b7ce1c6f963842f80a85f234b26ef6bf1b8da09
   languageName: node
   linkType: hard
```

ThatOneCalculator commented:
Well it'd be a bug in yarn v3, since it's using 3.2.0. Regardless, I'll investigate further.

ishowta commented:
> Well it'd be a bug in yarn v3, since it's using 3.2.0. Regardless, I'll investigate further.

https://github.com/yarnpkg/berry/issues/1142#issuecomment-607272809 これは行ってみましたか？あなたの環境のキャッシュに古いバージョンのキャッシュが残っているのではないでしょうか？

ThatOneCalculator commented:
@ishowta Tried that, didn't change anything.

acid-chicken commented:
How about

```js
		"browser-image-resizer": "misskey-dev/browser-image-resizer#tag=v2.2.1-misskey.2",
```

instead of

```js
		"browser-image-resizer": "git+https://github.com/misskey-dev/browser-image-resizer#v2.2.1-misskey.2",
``

?

ThatOneCalculator committed:
Change `browser-image-resizer`

ThatOneCalculator committed:
:finnadie:

ThatOneCalculator committed:
Merge branch 'misskey-dev:develop' into yarn-3

ThatOneCalculator committed:
Merge branch 'yarn-3' of https://github.com/ThatOneCalculator/misskey into yarn-3

ThatOneCalculator committed:
:goberserk:

ThatOneCalculator commented:
Why does it keep doing this :pensive: 

![image](https://user-images.githubusercontent.com/44733677/171090629-eef09ebc-4cbe-45b0-9378-8a99d5e303d2.png)

acid-chicken commented:
It seems `packages/*/.yarnrc.yml` exists even though it is not needed.

ThatOneCalculator committed:
:godmode:

ThatOneCalculator committed:
Remove yarnrc for workspaces

ThatOneCalculator committed:
🙏

ThatOneCalculator committed:
Add proper packages for eslint

ThatOneCalculator committed:
Skip extra install

ThatOneCalculator committed:
New steps for test.yml too

ThatOneCalculator commented:
I think it works now!! 

https://user-images.githubusercontent.com/44733677/171116299-1162b02e-9973-48c0-8fa7-45ada3cb5b08.mp4

ThatOneCalculator committed:
uhh

ThatOneCalculator commented:
@acid-chicken @ishowta Looks like lint now runs as expected!! (Fails on linting client, as normal)

Johann150 commented:
Would love if we can keep the lints for frontend and backend separate since the client lint is not fixed yet. Makes at least the backend lint meaningful.

ThatOneCalculator commented:
I'll see what I can do in the morning, it's 1 AM now 😴

rinsuki commented on .dockerignore:
```diff
@@ -3,6 +3,7 @@
 .travis
 .vscode
 .config
+.yarn
```
is this okay to exclude .yarn/plugins?

ThatOneCalculator committed:
Update .dockerignore

ThatOneCalculator commented on .github/workflows/lint.yml:
```diff
@@ -8,32 +8,20 @@ on:
-    - run: yarn install
-    - run: yarn --cwd ./packages/backend lint
+    - run: corepack enable
+    - run: yarn set version berry
```
`corepack enable` is local to the machine.

rinsuki replied:
when corepack is enabled, corepack automatically installs `packageManager` in package.json and use it. so you shouldn't need to `yarn set version berry` because `yarn` already points to yarn berry by corepack. if `yarn set version berry` is needed, I think something is wrong.

ThatOneCalculator committed:
Update lint.yml

ThatOneCalculator committed:
Update test.yml

ThatOneCalculator committed:
Merge branch 'yarn-3' of https://github.com/ThatOneCalculator/misskey into yarn-3

ThatOneCalculator commented:
Not sure why this is here, that file is removed :confused: 

![image](https://user-images.githubusercontent.com/44733677/171214898-bd3339b9-697b-48fd-b2ca-1f4d6fca372f.png)

ThatOneCalculator committed:
Make client/backend lints seperate workflows

ThatOneCalculator commented:
What is GitHub doing :raised_eyebrow: 

![image](https://user-images.githubusercontent.com/44733677/171216189-47a6ba92-f4ce-441b-af9c-6ba056e33ced.png)

![image](https://user-images.githubusercontent.com/44733677/171216140-19055e53-e5d3-4a11-bdd6-8c636abcb93d.png)

ishowta commented on package.json:
```diff
@@ -6,19 +6,24 @@
-		"build": "node ./scripts/build.js",
-		"start": "cd packages/backend && node --experimental-json-modules ./built/index.js",
-		"start:test": "cd packages/backend && cross-env NODE_ENV=test node --experimental-json-modules ./built/index.js",
+		"build": "yarn workspaces foreach run build",
```
gulpのビルドコマンドが実行されなさそうです

ishowta commented on package.json:
```diff
@@ -41,10 +47,14 @@
-		"@typescript-eslint/parser": "5.18.0",
+		"@typescript-eslint/eslint-plugin": "latest",
+		"@typescript-eslint/parser": "5.27.0",
 		"cross-env": "7.0.3",
-		"cypress": "9.5.3",
+		"cypress": "9.7.0",
+		"eslint-plugin-import": "^2.26.0",
+		"eslint-plugin-vue": "latest",
 		"start-server-and-test": "1.14.0",
-		"typescript": "4.6.3"
+		"typescript": "4.7.2",
+		"vue-eslint-parser": "^9.0.2"
```
`^` prefixやlatestの指定はmisskeyでは使われていないように見えます

ishowta commented on packages/backend/package.json:
```diff
@@ -9,17 +9,18 @@
 		"mocha": "cross-env NODE_ENV=test TS_NODE_FILES=true TS_NODE_TRANSPILE_ONLY=true TS_NODE_PROJECT=\"./test/tsconfig.json\" mocha",
 		"test": "npm run mocha"
 	},
+	"packageManager": "yarn@3.2.1",
```
```suggestion
```

ishowta replied:
個別のworkspaceにyarn.lockを置いたりpackageManagerを指定してはいけないと思います

ishowta replied:
また、workspace単位でのresolutionはできないようです

<img width="495" alt="スクリーンショット 2022-06-01 2 21 04" src="https://user-images.githubusercontent.com/7393198/171235170-9c22515c-e745-49b3-809c-2c994fd814ae.png">

ishowta commented on gulpfile.js:
```diff
@@ -16,11 +16,11 @@ gulp.task('copy:backend:views', () =>
 
 gulp.task('copy:client:fonts', () =>
-	gulp.src('./packages/client/node_modules/three/examples/fonts/**/*').pipe(gulp.dest('./built/_client_dist_/fonts/'))
+	gulp.src('./node_modules/three/examples/fonts/**/*').pipe(gulp.dest('./built/_client_dist_/fonts/'))
```
`nodeLinker: pnpm`が指定されている場合、`./node_modules/three`は存在しないと思います
おそらく殆どのnode_modulesのパスの変更を元に戻す必要があると思います :)

ThatOneCalculator replied:
Yep... :face_exhaling:

ishowta commented on .github/workflows/lint.yml:
```diff
@@ -8,32 +8,24 @@ on:
+    strategy:
+      matrix:
+        workspace:
+        - backend-771cdf
+        - client-66101c
```
ワークスペースごとにpackage.jsonに名前を設定するのが良いと思います

ThatOneCalculator replied:
@ishowta how do I give a name to a workspace? I can't seem to find documentation on how to do that...

acid-chicken replied:
just add the `name` field to package.json

ishowta commented on Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM node:18.0.0-alpine3.15 AS base
+FROM node:18.2.0-alpine3.15 AS base
```
Dockerでは正常に動作しましたか？

ThatOneCalculator committed:
package.json names

ThatOneCalculator committed:
pnpm = dont change paths

ThatOneCalculator committed:
Fix changelog

ThatOneCalculator committed:
:bowtie:

ThatOneCalculator commented:
Wtf?

![image](https://user-images.githubusercontent.com/44733677/171318440-78d14146-24ec-466b-bb1a-eca9b9c85b0d.png)

ThatOneCalculator committed:
I don't know why github workflows hates me...

ThatOneCalculator committed:
test

ThatOneCalculator committed:
?????

tamaina committed:
Merge branch 'develop' into pr/ThatOneCalculator/8764

tamaina commented:
Info: I merged develop branch.

tamaina commented:
browser-image-resizerのエラーはどうして発生するのだろう（あんまり乗り気ではないので調べる気がない）

tamaina commented:
(いくら乗り気でないとは言え、次のバージョンのMisskeyはメジャーバージョンをv13に上げるような気がするしその時にこのPRもマージされそうなので真面目にやる)

tamaina committed:
continue-on-error: true

tamaina commented:
browser-image-resizerの問題を解決しないとローカルでyarn buildできない

ishowta commented:
`yarn cache clean --all && YARN_CHECKSUM_BEHAVIOR=update yarn install --inline-builds | grep YN0000`

fetch step log
```
➤ YN0000: ┌ Fetch step
➤ YN0000: │ /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad STDOUT Packing browser-image-resizer@https://github.com/misskey-dev/browser-image-resizer.git#commit=a58834f5fe2af9f9f31ff115121aef3de6f9d416 from sources
➤ YN0000: │ /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad STDOUT Using Yarn Classic for bootstrap. Reason: "__metadata" key not found in yarn.lock, must be a Yarn classic lockfile
➤ YN0000: │ /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad STDOUT 
➤ YN0000: │ /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad STDOUT Resolving classic to a url...
➤ YN0000: │ /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad STDOUT Downloading https://github.com/yarnpkg/yarn/releases/download/v1.22.19/yarn-1.22.19.js...
➤ YN0000: │ /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad STDOUT Saving it into /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad/.yarn/releases/yarn-1.22.19.cjs...
➤ YN0000: │ /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad STDOUT Updating /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad/.yarnrc...
➤ YN0000: │ /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad STDOUT Done!
➤ YN0000: │ /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad STDOUT 
➤ YN0000: │ /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad STDOUT yarn install v1.22.19
➤ YN0000: │ /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad STDOUT [1/4] Resolving packages...
➤ YN0000: │ /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad STDOUT [2/4] Fetching packages...
➤ YN0000: │ /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad STDOUT [3/4] Linking dependencies...
➤ YN0000: │ /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad STDERR warning " > babel-loader@8.2.3" has unmet peer dependency "webpack@>=2".
➤ YN0000: │ /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad STDOUT [4/4] Building fresh packages...
➤ YN0000: │ /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad STDOUT Done in 4.16s.
➤ YN0000: │ /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad STDOUT 
➤ YN0000: │ /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad STDOUT yarn pack v1.22.19
➤ YN0000: │ /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad STDOUT success Wrote tarball to "/private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad/package.tgz".
➤ YN0000: │ /private/var/folders/kz/k7zs5pjj4ng49pjy6t52_5740000gp/T/xfs-478717ad STDOUT Done in 0.17s.
➤ YN0000: └ Completed in 55s 620ms
```

checksum

```diff
diff --git a/yarn.lock b/yarn.lock
index 27750d26e..5d2bf7a25 100644
--- a/yarn.lock
+++ b/yarn.lock
@@ -2859,7 +2859,7 @@ __metadata:
 "browser-image-resizer@misskey-dev/browser-image-resizer#tag=v2.2.1-misskey.2":
   version: 2.2.1-misskey.2
   resolution: "browser-image-resizer@https://github.com/misskey-dev/browser-image-resizer.git#commit=a58834f5fe2af9f9f31ff115121aef3de6f9d416"
-  checksum: 5a29181ee34a407b43069dcd5e3364c3066eb4af3ffdd262ba669ea37273ac232aadebff8204a6ae9ff8d74305429191feedb5d420449acf01e1bbf9759c4847
+  checksum: eb5ddfe7f6a2de96340ef420df9be03a75f2bfd8f568a60be22cc8f2ccdcb754105b7799cf706c09add3f9d82b7ce1c6f963842f80a85f234b26ef6bf1b8da09
   languageName: node
   linkType: hard
```

.yarn/cache/browser-image-resizer-***.zip

[browser-image-resizer-https-0fc11559f9-eb5ddfe7f6.zip](https://github.com/misskey-dev/misskey/files/8823797/browser-image-resizer-https-0fc11559f9-eb5ddfe7f6.zip)

tamaina commented:
（なぜチェックサムに齟齬が出たんだろうか）

acid-chicken commented:
Yarn classic が中で走ってる？

tamaina commented:
あとここらへんは消していいんじゃないかしら

![image](https://user-images.githubusercontent.com/7973572/171639969-277e6a1a-5eb0-4685-82dd-c82cefb5c65e.png)

acid-chicken commented:
It seems likely that this is an effect of the Git version or config.

tamaina committed:
Fix yarn.lock

Co-authored-by: ishowta <ishowta@gmail.com>

acid-chicken commented:
🤔 yarnpkg/berry#4152

ThatOneCalculator commented:
> I got exactly same checksum on GitHub Codespaces; maybe @ThatOneCalculator 's machine have a wrong cache?

Maybe, but i cleaned all packages/cache and installed/built again so I'm not sure why it would have the wrong cache

tamaina committed:
add @rollup/pluginutils

ishowta commented:
chartjs-adapter-date-fnsのpeerDependenciesにdate-fnsが存在しない (https://github.com/chartjs/chartjs-adapter-date-fns/issues/32)

依存関係にないパッケージに依存するのはデバッグが困難で良くないとされていて  (https://rushjs.io/pages/advanced/phantom_deps/) 、pnpmやyarnのpnpmモードではできないようになっている？

pnpmではルートに巻き上げるようにするオプションがあるが (https://pnpm.io/ja/6.x/npmrc#shamefully-hoist) 、yarn v3には無いかもしれない？

tamaina commented:
プルリクは放置されているので既に最悪な状況だと思う  
(chartjsの公式リポジトリっぽいのに…)

https://github.com/chartjs/chartjs-adapter-date-fns/pull/39

acid-chicken commented:
こういう場合は .yarnrc.yml に

```yaml
packageExtensions:
  "chartjs-adapter-date-fns@*":
    peerDependencies:
      date-fns: "*"
```

という感じに上書きしてやるワークアラウンドがある

tamaina committed:
add packageExtensions for chartjs_date-fns

Co-authored-by: acid-chicken <root@acid-chicken.com>

tamaina commented:
mochaがうごかない

ローカル

```
~/misskey$ yarn mocha
sh: 1: mocha: Permission denied
~/misskey$ sudo yarn mocha
sh: 1: mocha: Permission denied
```

GitHub Actions

```
# Run yarn mocha
  
sh: 1: mocha: not found
Error: Process completed with exit code 127.
```

acid-chicken commented:
そこら辺のスクリプト古いままだからそれはそう

tamaina commented:
あ～っ

tamaina committed:
npm run => yarn

acid-chicken commented on package.json:
```diff
@@ -6,30 +6,36 @@
 		"cy:open": "cypress open",
 		"cy:run": "cypress run",
 		"e2e": "start-server-and-test start:test http://localhost:61812 cy:run",
 		"mocha": "cd packages/backend && cross-env NODE_ENV=test TS_NODE_FILES=true TS_NODE_TRANSPILE_ONLY=true TS_NODE_PROJECT=\"./test/tsconfig.json\" npx mocha",
```
これを cross-env とか npx とか抜いて backend に移し、あとは `yarn workspace backend run mocha` とかで呼び出す

acid-chicken commented:
あと npx は yarn dlx になる

tamaina committed:
wip

tamaina committed:
yarn node => node

依存関係の記載不足の解消がかなり多いため、yarn nodeは使わない

tamaina commented:
依存関係の不足の解消が大変難しそうだったので`yarn node`は使わないようにした……

tamaina committed:
fix

tamaina committed:
fix http-signature

tamaina committed:
Merge branch 'develop' into pr/ThatOneCalculator/8764

tamaina committed:
fix

tamaina committed:
fix

acid-chicken commented:
テストが落ちるのってこういうこと？

```yaml
packageExtensions:
  "@bull-board/api@*":
    peerDependencies:
      "@bull-board/ui": "*"
```

felixmosh/bull-board#312

tamaina commented:
う～～～ん、やっぱ依存関係の不足をすべてpackageExtensionsで潰していかなければいけないのか

tamaina committed:
add packageExtensions

tamaina commented:
`/queue`にアクセスするとCannot find module 'ejs'になる

```
Error: Cannot find module 'ejs'
Require stack:
- /home/aqz/misskey/node_modules/.store/consolidate-virtual-50de78e82b/node_modules/consolidate/lib/consolidate.js
- /home/aqz/misskey/node_modules/.store/consolidate-virtual-50de78e82b/node_modules/consolidate/index.js
- /home/aqz/misskey/node_modules/.store/koa-views-virtual-bf8ad99819/node_modules/koa-views/src/index.js
```

tamaina committed:
:v:

tamaina committed:
remove `yarn set version berry`

tamaina committed:
yarn install --immutable

Co-authored-by: ishowta <ishowta@gmail.com>

tamaina committed:
https://github.com/misskey-dev/misskey/pull/8764#discussion_r885749892

tamaina commented on .github/workflows/lint.yml:
```diff
@@ -8,32 +8,24 @@ on:
+    - uses: actions/setup-node@v3.2.0
       with:
         node-version: 18.x
-        cache: 'yarn'
-        cache-dependency-path: |
```
https://dev.classmethod.jp/articles/caching-dependencies-in-workflow-execution-on-github-actions/

グローバルパッケージをキャッシュする(??)ものなのであんまり関係ない気がする

tamaina committed:
enable actions/setup-node's global cache

tamaina committed:
Update .gitignore

Co-authored-by: iwata <ishowta@gmail.com>

tamaina committed:
revival gulp build
https://github.com/misskey-dev/misskey/pull/8764/files/66ab7591bff9b35255219057e33399a06260aa31#r885899944

tamaina committed:
fix lockfile

ThatOneCalculator commented:
Ty Tamania for all the fixes :)

Is there anything more that needs to be done?

tamaina commented:
packageExtensionsの記載が必要なパッケージが洗い出せているわけではないので、テストが完全に動くまではマージしたくない……

https://github.com/misskey-dev/misskey/issues/8782

rinsuki commented:
> Also, since there are opinions like https://github.com/misskey-dev/misskey/issues/5858#issuecomment-611626477, it would be better if Corepack could be used as a boot loader for Yarn berry (instead of placing .yarn/releases/yarn-3.2.1.cjs directly).
> https://github.com/misskey-dev/misskey/pull/8764#issuecomment-1140545858

FYI, in my latest opinion, yarn berry requiring puts big minified js of all plugins to repository, so I think one more big minified js is acceptable for rescues anyone that not/forgot/hard to enable corepack.

rinsuki commented:
> packageExtensionsの記載が必要なパッケージを完全に洗い出せているわけではないので、テストが完全に動くまではマージしたくない……

そもそもテストってちゃんと全部網羅してるんだっけ？みたいな話も………

あとなんで pnpm モード使いたいんだっけ

tamaina committed:
Merge branch 'develop' into pr/ThatOneCalculator/8764

tamaina committed:
Update packages/backend/package.json

Co-authored-by: iwata <ishowta@gmail.com>

tamaina committed:
remove packagemanager

tamaina committed:
Merge branch 'yarn-3' of https://github.com/thatonecalculator/misskey into pr/ThatOneCalculator/8764

tamaina commented:
p1.a9z.devで適用中

ThatOneCalculator committed:
Merge branch 'misskey-dev:develop' into yarn-3

ThatOneCalculator committed:
`yarn run gulp` in `build` instead of just `gulp`

ThatOneCalculator commented:
Made [308f57d](https://github.com/misskey-dev/misskey/pull/8764/commits/308f57d18d0302b364a9234eb721ca09e5de245b) as I tested this PR on https://stop.voring.me and I got a command not found error due to gulp not being installed globally.

ThatOneCalculator committed:
Update CHANGELOG.md

ThatOneCalculator commented:
I've been running this PR for about a week, no issues. Requesting a review from @syuilo

ThatOneCalculator commented:
My account is back as is this PR

ThatOneCalculator commented:
uh oh

ThatOneCalculator commented:
Looks like I'm gonna have to fix the lockfile, yippee...

ThatOneCalculator commented:
@tamaina I'm having trouble updating the lockfile with the merge conflicts, do you think you could help? :pray:

tamaina committed:
Merge branch 'develop' into pr/ThatOneCalculator/8764

tamaina committed:
update lockfile

tamaina commented:
merged

tamaina commented:
developにマージしたわけではない

ThatOneCalculator commented:
Tested and working on https://stop.voring.me

ThatOneCalculator committed:
Merge branch 'misskey-dev:develop' into yarn-3

ThatOneCalculator committed:
Merge branch 'yarn-3' of https://github.com/ThatOneCalculator/misskey into yarn-3

ThatOneCalculator committed:
Add .yarn/cache to gitignore for packages

ThatOneCalculator commented:
@tamaina given deps have been updated again, should we update the lockfiles here too?

tamaina commented:
Conflicts must be resolved.

tamaina commented:
@syuilo 結局どう考える？ (p1.a9z.devで動かしているけど特に問題はなさそう)

syuilo commented:
じっくり見てないけど良いとは思う
これユーザー側でyarnの更新必要になる？

acid-chicken commented:
> これユーザー側でyarnの更新必要になる？

Yarn の更新というより Corepack のオプトインがいる。オプトインしたらローカルに自分で Yarn を入れるのはむしろ非推奨になる。

tamaina commented:
> これユーザー側でyarnの更新必要になる？

はい

1. yarnをインストールしている場合はアンインストール
2. (sudo?) corepack enable yarn
3. npx yarnは使わない

ThatOneCalculator commented:
Sudo isn't needed for corepack :)

tamaina committed:
Merge branch 'develop' into pr/ThatOneCalculator/8764

tamaina committed:
?

tamaina committed:
move resolutions

tamaina committed:
Merge branch 'develop' into pr/ThatOneCalculator/8764

tamaina committed:
:v:

tamaina committed:
Merge branch 'develop' into pr/ThatOneCalculator/8764

tamaina committed:
update lockfie

tamaina committed:
Merge branch 'develop' into pr/ThatOneCalculator/8764

sousuke0422 commented:
`@tensorflow/tfjs`を明確にpackage.jsonに入れないと起動できなさそう

tamaina committed:
Merge branch 'develop' into pr/ThatOneCalculator/8764

tamaina committed:
update-lockfile

tamaina committed:
Merge branch 'develop' into pr/ThatOneCalculator/8764

tamaina committed:
update yarn.lock

tamaina committed:
fix build

tamaina committed:
fix

tamaina committed:
fix

tamaina committed:
fix: add @tensorflow/tfjs

tamaina committed:
ダメ

tamaina commented:
なんかエラー出た

![image](https://user-images.githubusercontent.com/7973572/191164747-9276843d-0ed9-4463-9c44-21575342a194.png)

```
node:internal/modules/cjs/loader:956
  const err = new Error(message);
              ^

Error: Cannot find module 'long'
Require stack:
- /home/ubuntu/misskey-dev/node_modules/.store/@tensorflow-tfjs-npm-3.20.0-3984b7fecc/node_modules/@tensorflow/tfjs/dist/tf.node.js
- /home/ubuntu/misskey-dev/node_modules/.store/nsfwjs-virtual-fe284db5a1/node_modules/nsfwjs/dist/index.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:956:15)
    at Module._load (node:internal/modules/cjs/loader:804:27)
    at Module.require (node:internal/modules/cjs/loader:1022:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/home/ubuntu/misskey-dev/node_modules/.store/@tensorflow-tfjs-npm-3.20.0-3984b7fecc/node_modules/@tensorflow/tfjs/dist/tf.node.js:24:19)
    at Module._compile (node:internal/modules/cjs/loader:1120:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1174:10)
    at Module.load (node:internal/modules/cjs/loader:998:32)
    at Module._load (node:internal/modules/cjs/loader:839:12)
    at Module.require (node:internal/modules/cjs/loader:1022:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/ubuntu/misskey-dev/node_modules/.store/@tensorflow-tfjs-npm-3.20.0-3984b7fecc/node_modules/@tensorflow/tfjs/dist/tf.node.js',
    '/home/ubuntu/misskey-dev/node_modules/.store/nsfwjs-virtual-fe284db5a1/node_modules/nsfwjs/dist/index.js'
  ]
}

Node.js v18.7.0
```

tamaina committed:
modify lockfile

tamaina commented:
原因これじゃね説

https://github.com/tensorflow/tfjs/issues/6841

tamaina commented:
developに追従した

syuilo commented:
yosasou

tamaina commented:
いいっちゃいいんだけど、いまマージするとlongのことを忘れられてあとで面倒事が起きる未来しか見えない

rinsuki commented:
lockfileいじるんじゃなくてもっといい方法ないっけ .yarnrc.yml に書くとかだった気がする

acid-chicken commented:
再掲

> こういう場合は .yarnrc.yml に
> 
> ```yaml
> packageExtensions:
>   "chartjs-adapter-date-fns@*":
>     peerDependencies:
>       date-fns: "*"
> ```
> 
> という感じに上書きしてやるワークアラウンドがある

tamaina committed:
use yarnrc

tamaina committed:
Merge branch 'develop' into pr/ThatOneCalculator/8764

tamaina commented:
.yarnrc.ymlかえた

tamaina committed:
Merge branch 'develop' into pr/ThatOneCalculator/8764

tamaina committed:
update yarnlock

tamaina committed:
add jest-mock

tamaina committed:
Merge branch 'develop' into pr/ThatOneCalculator/8764

tamaina committed:
update-lockfile

tamaina committed:
Merge branch 'develop' into pr/ThatOneCalculator/8764

tamaina committed:
update lockfile

tamaina commented:
developに追従しあ

tamaina committed:
Merge branch 'develop' into pr/ThatOneCalculator/8764

tamaina commented:
マージしちゃう？

tamaina committed:
Merge branch 'develop' into pr/ThatOneCalculator/8764

tamaina commented:
yarn v3はメモリを大量に使用するようですが、何か解決策はあるでしょうか？

![image](https://user-images.githubusercontent.com/7973572/201521900-08a53f10-4774-41cb-8e2c-f7025824a90e.png)

tamaina committed:
update lockfile

tamaina commented:
v1では1536MiB程度のVMでビルドできるのでちょっと微妙かなと

tamaina commented:
yarn install --immutableであれば1GiBで足りるかも

rinsuki commented:
そもそもproductionのビルドってv1ブランチでも1GBで足りるんだっけ？vite入ったから行ける？

tamaina commented:
無理だった
でも1536MiBならいけた

tamaina committed:
Merge branch 'develop' into pr/ThatOneCalculator/8764

syuilo commented on packages/backend/package.json:
```diff
@@ -1,28 +1,32 @@
 	"type": "module",
 	"scripts": {
+		"start": "node --experimental-json-modules ./built/index.js",
+		"start:test": "NODE_ENV=test node --experimental-json-modules ./built/index.js",
```
nodeのバージョン上げたから --experimental-json-modules は要らなくなったはず

syuilo commented on packages/client/package.json:
```diff
@@ -76,6 +74,8 @@
 		"@types/tinycolor2": "1.4.3",
 		"@types/uuid": "8.3.4",
+		"@types/websocket": "1.0.5",
+		"@types/ws": "8.5.3",
```
これ追加したのはなぜ？

tamaina replied:
動かなかったからじゃないかしら

tamaina replied:
んあ、別PRで提供せよという意味かしら

syuilo replied:
developでは無くても動いてると思うけど、このPRによってそれがないと動かなくなった？

tamaina replied:
(動かないというか型支援が聞かない

developで型支援できているのだとしたらなぜdevelopで可能になっているかが謎かも

syuilo commented on packages/backend/package.json:
```diff
@@ -53,11 +57,13 @@
 		"ioredis": "4.28.5",
 		"ip-cidr": "3.0.10",
 		"is-svg": "4.3.2",
+		"jest-mock": "^29.0.3",
```
要る？
要るとしてもdevDependenciesじゃないかしら

syuilo commented on packages/backend/package.json:
```diff
@@ -82,6 +88,7 @@
 		"node-fetch": "3.3.0",
 		"nodemailer": "6.8.0",
 		"nsfwjs": "2.4.2",
+		"oauth": "^0.10.0",
```
要る？

tamaina replied:
DiscordServerServiceとGithubServerServiceでは使われているけど

tamaina committed:
Merge branch 'develop' into pr/ThatOneCalculator/8764

tamaina committed:
move jest-mock, update lockfile

tamaina commented:
Fastifyと統合するのちょっと骨折れそう

syuilo reviewed (approved):
YOSASOU

syuilo commented:
🙏🏻🙏🏻🙏🏻

ThatOneCalculator commented:
:tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada:

syuilo commented:
nodeLinker: pnpmだと型の不具合が起きるっぽい
https://mstdn.rinsuki.net/@rinsuki/109556531278084299

rinsuki commented:
いやこれ peerDependencies に追加すれば解決する
